### PR TITLE
Fix Issue 24083 - Int128.opCmp's behavior with negative numbers is inconsistent with Int128.opEquals

### PR DESCRIPTION
**Pull request by n8sh**
Needed to make opCmp a template to avoid causing `Int128(-1L) < ulong.max` to return the wrong value or causing the compiler to complain that it can't determine which overload to call for `Int128(1) > 10`.

### CODE REVIEW

File: std/int128.d [hunk 1]
```diff
@@ -263,13 +263,17 @@ public struct Int128
...
     }
 
-    /** support signed arithmentic comparison operators < <= > >=
+    /** support arithmentic comparison operators < <= > >=
      * Params: op2 = right hand operand
      * Returns: -1 for less than, 0 for equals, 1 for greater than
      */
-    int opCmp(long op2) const
+    int opCmp(Long)(const Long op2) const
+    if (__traits(isIntegral, Long) && is(Long : long))
```
Comment (n8sh) [inline std/int128.d:271]: The `&& is(Long : long)` excludes vectors and would exclude cent & ucent if those were implemented.

Minus the exclusion of cent/ucent this could also be written importing `std.traits` as `if (isIntegral!Long || isSomeChar!Long || isBoolean!Long)`.

**Comment by n8sh** [comment]
In a perfect world opCmp would look like this:

```d
int opCmp(Int128 rhs)
{
   // ...
}

int opCmp(long rhs)
{
    return opCmp(Int128(rhs >> 63, rhs));
}

int opCmp(ulong rhs)
{
    return opCmp(Int128(0, rhs));
}
```

But this wouldn't work with `Int128(5, 6) > 10` (which is in the unittest) because the compiler can't decide which overload to use.

File: std/int128.d [hunk 2]
```diff
@@ -263,13 +263,19 @@ public struct Int128
         return this == op2 ? 0 : gt(this.data, op2.data) * 2 - 1;
     }
 
-    /** support signed arithmentic comparison operators < <= > >=
```
Comment (RazvanN7) [inline std/int128.d:266]: Why is this deleted? Also, the previous overload still contains "signed" in its documentation.

File: std/int128.d [hunk 3]
```diff
@@ -263,13 +263,19 @@ public struct Int128
...
      * Params: op2 = right hand operand
      * Returns: -1 for less than, 0 for equals, 1 for greater than
      */
-    int opCmp(long op2) const
+    int opCmp(Long)(const Long op2) const
+    if (is(Long : long))
     {
-        return opCmp(Int128(0, op2));
+        static if (!__traits(isIntegral, Long)) // Long converts to integer through alias this.
+            return opCmp(__traits(getMember, op2, __traits(getAliasThis, Long)[0]));
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply std/int128.d:274]: I think it's covered by the template constraint `is(Long : long)`.

[the remark this answers is not in the document]
Comment (dkorpel) [reply std/int128.d:274]: My bigger concern is whether we need to add code to support `alias this` in the first place. It reeks a bit of Phobos' generality creep. It's also not part of the issue, so it could use its own separate PR.

File: std/int128.d (hunk 2 above)
Comment (n8sh) [reply std/int128.d:266]: Because now this is a template function that handles both signed and unsigned comparison. The previous overload is for Int128 only so it's still signed.

File: std/int128.d [hunk 4]
```diff
@@ -263,13 +263,19 @@ public struct Int128
...
+    /** support arithmentic comparison operators < <= > >=
      * Params: op2 = right hand operand
      * Returns: -1 for less than, 0 for equals, 1 for greater than
      */
-    int opCmp(long op2) const
+    int opCmp(Long)(const Long op2) const
+    if (is(Long : long))
     {
-        return opCmp(Int128(0, op2));
+        static if (!__traits(isIntegral, Long)) // Long converts to integer through alias this.
```
[the remark this answers is not in the document]
Comment (n8sh) [reply std/int128.d:273]: Yes I'll change the wording.

File: std/int128.d (hunk 3 above)
[the remark this answers is not in the document]
Comment (n8sh) [reply std/int128.d:274]: [That is to say, this code currently works so breaking it would be a regression:](https://run.dlang.io/is/BVoXVA)

```d
struct Wrapper
{
    long value;;
    alias value this;
}
void main()
{
    import std.int128 : Int128;
    assert(Int128(1L) < Wrapper(2L));
}
```

**Review by dkorpel — APPROVED** [review]

**Comment by n8sh** [comment]
> std/int128.d(411:19)[warn]: Variable w is never modified and could have been declared const or immutable.
> make: *** [posix.mak:587: dscanner] Error 1

As D-Scanner's blindness to function calls with omitted parentheses is a known issue (https://github.com/dlang-community/D-Scanner/issues/325) I question whether this style rule should be enforced in Phobos.
